cmake/pcl_targets.cmake

```diff
@@ -411,6 +411,45 @@ macro(PCL_ADD_TEST _name _exename)
   add_dependencies(tests ${_exename})
 endmacro()
 
+###############################################################################
+# Add a benchmark target.
+# _name The benchmark name.
+# _exename The exe name.
+# ARGN :
+#    ARGUMENTS Arguments for benchmark executable
+#    LINK_WITH link benchmark executable with libraries
+macro(PCL_ADD_BENCHMARK _name)
+  set(options)
+  set(oneValueArgs)
+  set(multiValueArgs ARGUMENTS LINK_WITH)
+  cmake_parse_arguments(PCL_ADD_BENCHMARK "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN})
+
+  add_executable(benchmark_${_name} ${_name}.cpp)
+  set_target_properties(benchmark_${_name} PROPERTIES FOLDER "Benchmarks")
+  target_link_libraries(benchmark_${_name} benchmark::benchmark ${PCL_ADD_BENCHMARK_LINK_WITH})
+  set_target_properties(benchmark_${_name} PROPERTIES RUNTIME_OUTPUT_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR})
+  
+  #Only applies to MSVC
+  if(MSVC)
+    #Requires CMAKE version 3.13.0
+    if(CMAKE_VERSION VERSION_LESS "3.13.0" AND (NOT ArgumentWarningShown))
```

Comment:
`ArgumentWarningShown` will not work as excepted currently. The variable will be expanded within scope of `benchmarks/CMakeLists.txt` , so it is only available there (and in sub directories of it). So when calling `PCL_ADD_BENCHMARK` first wihin a subdirectory and the in the main directory, the variable is unset at first time. So you should use a [global variable](https://stackoverflow.com/questions/10031953/how-to-set-the-global-variable-in-a-function-for-cmake) instead for it (I don't recommend `CACHE INTERNAL` here, as then the warning will be only shown when the CMakeCache will be created and not anytime `configure` will be called).

Reply:
Well, it works if its a macro, so its inserted in `benchmarks/CMakeLists.txt` in which the parent scope will be our main CMakeLists?
Its at least the same solution I used for the unit tests and I'm pretty sure it worked as expected, but my memory might fail me here.

However, if its changed to a function, we probably should change it to a global variable.

Reply:
> Well, it works if its a macro, so its inserted in `benchmarks/CMakeLists.txt` in which the parent scope will be our main CMakeLists?

No, this our main CMakeLists will not see this variable.

Time for a short explanation.

### Example 1 ###

Let's asume we have following CMake files:

benchmark/CMakeList.txt
```cmake
PCL_ADD_BENCHMARK(features_normal_3d FILES ...)

add_subdirectory(another_dir_with_benachmark)
add_subdirectory(another_dir_with_benachmark2)
```

benchmark/another_dir_with_benachmark/CMakeList.txt
```cmake
PCL_ADD_BENCHMARK(another_benchmark FILES ...)
```

```
benchmark/another_dir_with_benachmark2/CMakeList.txt
```cmake
PCL_ADD_BENCHMARK(another_benchmark2 FILES ...)
```

This works as you want

### Example 1 - Explanation ###

benchmark/CMakeList.txt
```cmake
# ArgumentWarningShown is not set until now => PCL_ADD_BENCHMARK will show a warning
PCL_ADD_BENCHMARK(features_normal_3d FILES ...)
# ArgumentWarningShown is now TRUE after calling PCL_ADD_BENCHMARK
add_subdirectory(another_dir_with_benachmark)
add_subdirectory(another_dir_with_benachmark2)
```

benchmark/another_dir_with_benachmark/CMakeList.txt
```cmake
# ArgumentWarningShown is TRUE like in parent CMakeLists => no warning
PCL_ADD_BENCHMARK(another_benchmark FILES ...)
#
```

```
benchmark/another_dir_with_benachmark2/CMakeList.txt
```cmake
# ArgumentWarningShown is TRUE like in parent CMakeLists => no warning
PCL_ADD_BENCHMARK(another_benchmark2 FILES ...)
```

### Example 2 ###

Now let us a modify the snippet a little bit, by moving first `PCL_ADD_BENCHMARK` below both calls to `add_subdirectory` :

benchmark/CMakeList.txt
```cmake
add_subdirectory(another_dir_with_benachmark)
add_subdirectory(another_dir_with_benachmark2)

PCL_ADD_BENCHMARK(features_normal_3d FILES ...)
```

All other files are equal

And now you are getting the warning 3 times. Why? See below

### Example 2 - Explanation ###

benchmark/CMakeList.txt
```cmake
# ArgumentWarningShown is not set until now
add_subdirectory(another_dir_with_benachmark)
add_subdirectory(another_dir_with_benachmark2)

# ArgumentWarningShown is not set until now, as changes from child scope are not injected into parent scope => PCL_ADD_BENCHMARK will show a warning
PCL_ADD_BENCHMARK(features_normal_3d FILES ...)
```

benchmark/another_dir_with_benachmark/CMakeList.txt
```cmake
# ArgumentWarningShown is not set until now => PCL_ADD_BENCHMARK will show a warning
PCL_ADD_BENCHMARK(another_benchmark FILES ...)
# ArgumentWarningShown is now TRUE after calling PCL_ADD_BENCHMARK
#
```

```
benchmark/another_dir_with_benachmark2/CMakeList.txt
```cmake
# ArgumentWarningShown is not set until now => PCL_ADD_BENCHMARK will show a warning
PCL_ADD_BENCHMARK(another_benchmark2 FILES ...)
# ArgumentWarningShown is now TRUE after calling PCL_ADD_BENCHMARK
```

### Conclusion ###
Calling `add_subdirectory` is like passing all existing variables to a  C++ function as copy and not as reference. The caller will not see the changes of the callee.

Nevertheless it is possible in CMake to modify parent scope:
- via global states
- using `PARENT_SCOPE`

`PARENT_SCOPE` will be usually used in functions, where you pass the variable name and want that the function stores the result into this variable. In other cases I don't recommend using `PARENT_SCOPE`. We had it in our framework a long time, but it was bug prune, as in case someone is missing one time a call to `PARENT_SCOPE`, you don't first see why sth. is not working like excepted.

As the variable `ArgumentWarningShown` is only related to benchmark I would use:
```cmake
set_target_properties(ArgumentWarningShown run_benchmark PCL_ARGUMENTS_WARNING_SHOWN)
if(NOT ArgumentWarningShown)
  ..
  set_target_properties(run_benchmark PROPERTIES PCL_ARGUMENTS_WARNING_SHOWN TRUE)
endif()
```

Reply:
Yes, thats true if its organized differently as first intended, it will stop working. Thanks for the thorough explanation.
@mvieth can you reorganize appropriately?
And maybe change the name `ArgumentWarningShown `to `BenchArgumentWarningShown` so it doesn't clash with the one in Unit Tests.

Reply:
So, macros are basically evil in almost all languages. (:crab:)

Reply:
@SunBlack I assume you mean `get_target_property` in the first line, and run_benchmark**s**, or did I misunderstand something?